scapy/contrib/wireguard.py

```diff
@@ -0,0 +1,119 @@
+# This file is part of Scapy
+# Scapy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# any later version.
+#
+# Scapy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Scapy. If not, see <http://www.gnu.org/licenses/>.
+
+
+# scapy.contrib.description = WireGuard
+# scapy.contrib.status = loads
+
+"""WireGuard Module
+Implements the WireGuard network tunnel protocol.
+Based on the whitepaper: https://www.wireguard.com/papers/wireguard.pdf
+"""
+
+from scapy.fields import ByteEnumField, ThreeBytesField, XIntField, \
+    XStrFixedLenField, LongField, XStrField
+from scapy.layers.inet import UDP
+from scapy.packet import Packet, bind_layers
+
+
+class Wireguard(Packet):
+    """
+    Wrapper that only contains the message type.
+    """
+    name = "Wireguard"
+
+    fields_desc = [
+        ByteEnumField(
+            "type", 1,
+            {
+                1: "initiate",
+                2: "respond",
+                3: "cookie reply",
+                4: "transport"
+            }
+        ),
+        ThreeBytesField("reserved", 0)
+    ]
+
+    def guess_payload_class(self, payload):
```

Comment:
Instead of using `guess_paylod_class` + `overloaded_fields`, you should add a `bind_layers` call for each. It's much more

> minimalistic by design

:smile: